### ISSUE TITLE
version index limited to 16 characters, indexing messes up silently.

### ISSUE THREAD

**Issue by kentfredric** [andk/pause#191]
As exemplified by Net::FullAuto, as soon as you exceed 16 characters, new releases will start getting indexed at a truncated version.

This is also imagined to happen with very large integer values, which might result in strange behaviour as  "1234" becomes "123" ( except with 16 instead of 3 )
```
 $ wget -qO- cpan.metacpan.org/modules/02packages.details.txt.gz | zgrep Net::FullAuto
Net::FullAuto                  0.99999999999999  R/RE/REEDFISH/Net-FullAuto-0.9999999999999905.tar.gz
Net::FullAuto::FA_Core            undef  R/RE/REEDFISH/Net-FullAuto-0.9999999999999905.tar.gz
Net::FullAuto::FA_DB              undef  R/RE/REEDFISH/Net-FullAuto-0.9999999999999905.tar.gz
Net::FullAuto::Getline            undef  R/RE/REEDFISH/Net-FullAuto-0.9999999999999905.tar.gz
Net::FullAuto::MemoryHandle       undef  R/RE/REEDFISH/Net-FullAuto-0.9999999999999905.tar.gz
```

And the last 5 releases have been similarly indexed, and indexed with the truncated version:
```
 wget -qO- http://cpanmetadb.plackperl.org/v1.0/history/Net::FullAuto | grep '0.99999999999999'
Net::FullAuto                  0.99999999999999  R/RE/REEDFISH/Net-FullAuto-0.99999999999999.tar.gz
Net::FullAuto                  0.99999999999999  R/RE/REEDFISH/Net-FullAuto-0.9999999999999901.tar.gz
Net::FullAuto                  0.99999999999999  R/RE/REEDFISH/Net-FullAuto-0.9999999999999902.tar.gz
Net::FullAuto                  0.99999999999999  R/RE/REEDFISH/Net-FullAuto-0.9999999999999903.tar.gz
Net::FullAuto                  0.99999999999999  R/RE/REEDFISH/Net-FullAuto-0.9999999999999904.tar.gz
```

**Comment by karenetheridge** [comment]
I don't really see a problem with this, so long as the PAUSE index email receipt indicates what the indexed version was.  After all, we have to have a limit *at some point* if we expect versions to be evaluated in numeric context.

Especially given that Net::FullAuto is the first and perhaps only distribution to ever encounter this situation, I'd classify it as "not a bug -- works as intended".

**Comment by kentfredric** [comment]
In this case, uploading new content and indexing it at a truncated version is an error, regardless. 

Its not something a user requested, and its not something a user expects.

Overriding the index value ( that is, assigning a new result for the key "0." . ("9" x 14)  ) is better than nothing, because it at least allows the distribution to be installed at the new version.

Given that standard PAUSE operating procedure is to refuse to reindex an existing version to a new tarball, that exceeding the truncation length implicitly permits this behaviour is a clear violation of design.

Thus, in order to be consistent with other versions, if a version exceeds truncation limit and thus emulates the "new tarball, same version" scenario, that it should behave the same.

- Send an email to the author telling them indexing failed because the versions didn't increase
- Refuse to index
- Be overrideable via poking into the PAUSE website telling them "no, I didn't just make a mistake I'm fine with the indexed version not incrementing, I don't care that nobody on the planet will be able to upgrade from one of my indexed versions to another one because they all look the same to the CPAN client"

**Comment by ap** [comment]
> I don't really see a problem with this

You can’t be serious.

> we have to have a limit *at some point*

So you just index it that way, I see… the MySQL school of data integrity. “Relax man, your string was long so I just kept as much as I could fit in your VARCHAR(30) and I guess the rest went on the floor… why raise an error about that? Just chill.”

*Of course* you have to limit at some point. Nobody said or thought otherwise. But what ought to happen when you do is you raise an error, not just blithely proceed. PAUSE should refuse to index the distribution if it can’t faithfully represent its version.

**Comment by ap** [comment]
> Thus, in order to be consistent with other versions, if a version exceeds truncation limit and thus emulates the "new tarball, same version" scenario, that it should behave the same.

Disagree. Semantically what’s going on is that the version is too long to represent, so that should be detected and PAUSE should just throw up its hands. The fact that if you cut off the version at the maximum length you get some sort of value is an accident of implementation that shouldn’t carry weight in the design, rather than meaningful data you should then try to reason about.

**Comment by karenetheridge** [comment]
> Given that standard PAUSE operating procedure is to refuse to reindex an existing version to a new tarball

Not so!  I can upload Foo-Bar-0.001.tar.gz containing $VERSION = '0.001', and then upload Foo-Bar-0.001b.tar.gz containing the same $VERSION and now my new tarball is indexed. This is periodically exploited intentionally to repair a botched tarball upload.

> that exceeding the truncation length implicitly permits this behaviour is a clear violation of design.

Because of the above, I argue that this is consistent behaviour.

**Comment by zoffixznet** [comment]
> Its not something a user requested, and its not something a user expects.

Just because the author of Net::FullAuto spams uploads several times a day to get attention to his [malware](http://cpanratings.perl.org/dist/Net-FullAuto) is not a valid justification to expend volunteer effort restructuring PAUSE.

Since it's possible to upload same-version dist using different tarballs, this issue should be closed until a real user-case emerges.

**Comment by ap** [comment]
> Just because the author of Net::FullAuto spams uploads several times a day to get attention to his malware is not a valid justification to expend volunteer effort restructuring PAUSE.

First, there is no such thing as expending volunteer effort; that’s why it’s called volunteer effort. Either someone makes the effort voluntarily or they don’t. You don’t get to allocate resources or demand justification, unless the effort is no longer voluntary.

But that aside, just because it was NFA that highlighted the issue doesn’t mean the issue shouldn’t be fixed for everyone else.

Not that anybody is even suggesting a change that would actually help the author of NFA (such as increasing the version length limit) – the serious proposals all affect him adversely.

**Comment by zoffixznet** [comment]
> First, there is no such thing as expending volunteer effort; that’s why it’s called volunteer effort. Either someone makes the effort voluntarily or they don’t. You don’t get to allocate resources or demand justification, unless the effort is no longer voluntary.

Opening an Issue on a project is a request to expend volunteer effort to fix it.

> But that aside, just because it was NFA that highlighted the issue doesn’t mean the issue shouldn’t be fixed for everyone else.

At the moment, I don't see any issues anyone else is having or is expected to have any time in the future. If you feel so strongly about it, I invite you to submit a PR to rectify it.

**Comment by ap** [comment]
> Opening an Issue on a project is a request to expend volunteer effort to fix it.

Right, a request. You were talking valid this and justification that, so I wasn’t sure you were keeping that straight.

**Comment by kentfredric** [comment]
> Opening an Issue on a project is a request to expend volunteer effort to fix it.

Actually, not in this case.

I was quite happy to just tell NFA that they'd run out of versions, and that they would need to buy a better version scheme to continue playing.

However, as many as _3_ people immediately decided that this behavior was much better classified as a bug, and at least one of those people has commit access from time to time to PAUSE (probably more, I'm just being conservative ), and that person expressed as such in a context of *intent to fix*

```
03:54:07 <@rjbs> please file a bug and I'll look at fixing it soonish
```

So no "request for volunteers" is necessary, as in idle conversation, a volunteer piped up and appointed themselves.

**Comment by ranguard** [comment]
+1 for this being changed... I'm trying to upload P6 modules, and not get in people's way when they want to do so them selves, so I'm using: version 0.000.000_YYMDDhhmmss eg: Acme-Meow-0.000.000_151220210845.tar.gz - I know I'm hitting a p6 issue in the indexing, which  https://github.com/andk/pause/pull/193 will fix, but thought I'd mention my weird use case.

**Comment by rjbs** [comment]
This is a bug and I intend to submit a fix.

**Comment by zoffixznet** [comment]
@rjbs mentioning just in case you didn't read every message on this Issue: @ranguard's usecase above requires 21 chars for the version number. Not sure if you were planning to extend the max length or just generate an error email when the current max length of 16 chars is hit.

**Comment by rjbs** [comment]
This problem is fixed by #195.

I have no plans to extend the maximum length, but Andreas can do so it he thinks it's worth the trouble.  This should not affect @ranguard, because p6 data is stored in its own tables.  `p6dists.ver` is a char(255).  (Why char and not varchar? I have no idea and do not plan to investigate!)

**Comment by rjbs** [comment]
I am closing this, as #195 has been merged.  Thanks, @andk!
